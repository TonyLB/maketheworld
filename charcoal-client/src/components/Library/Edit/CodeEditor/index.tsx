import { FunctionComponent, useCallback, useMemo, useState } from "react";
import { createEditor, Descendant } from "slate";
import { withHistory } from "slate-history";
import { Editable, Slate, withReact } from "slate-react";
import stringToSlate, { slateToString } from "./stringToSlate";

type CodeEditorProps = {
    source: string;
    onChange: (value: string) => void;
}

const Leaf = ({ attributes, children, leaf }: { attributes: any, children: any, leaf: any }) => {
    return (
        <span
            {...attributes}
            style={{ backgroundColor: leaf.error && '#ffeeba' }}
            spellCheck={false}
        >
            {children}
        </span>
    )
}

export const CodeEditor: FunctionComponent<CodeEditorProps> = ({ source, onChange }) => {
    const [editor] = useState(() => withHistory(withReact(createEditor())))
    const value = useMemo(() => (stringToSlate(source)), [source])
    const onChangeHandler = useCallback((nodes: Descendant[]) => {
        onChange(slateToString(nodes))
    }, [onChange])
    const renderLeaf = useCallback(props => (<Leaf { ...props } />), [])
    return <Slate editor={editor} value={value} onChange={onChangeHandler}>
        <Editable
            renderLeaf={renderLeaf}
        />
    </Slate>
}

export default CodeEditor
