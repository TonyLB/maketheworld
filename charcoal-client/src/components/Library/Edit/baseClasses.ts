import { BaseEditor, Selection } from 'slate'
import { ReactEditor } from 'slate-react'

export type CustomText = {
    text: string;
}

export type CustomLineElement = {
    type: 'line';
    children: CustomText[]
}

export type CustomActionLinkElement = {
    type: 'actionLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomFeatureLinkElement = {
    type: 'featureLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomLinkElement = CustomActionLinkElement | CustomFeatureLinkElement

export type CustomLineBreak = {
    type: 'lineBreak';
}

export type CustomParagraphContents = CustomText | CustomActionLinkElement | CustomFeatureLinkElement | CustomLineBreak

export const isCustomLineBreak = (item: CustomParagraphContents): item is CustomLineBreak => ('type' in item && item.type === 'lineBreak')
export const isCustomActionLink = (item: CustomParagraphContents): item is CustomActionLinkElement => ('type' in item && item.type === 'actionLink')
export const isCustomFeatureLink = (item: CustomParagraphContents): item is CustomFeatureLinkElement => ('type' in item && item.type === 'featureLink')
export const isCustomLink = (item: CustomParagraphContents): item is CustomLinkElement => (isCustomActionLink(item) || isCustomFeatureLink(item))
export const isCustomText = (item: CustomParagraphContents): item is CustomText => ('text' in item)

export type CustomParagraphElement = {
    type: 'paragraph';
    children: CustomParagraphContents[]
}

export type CustomDescriptionElement = {
    type: 'description';
    children: CustomParagraphElement[]
}

type CustomElement = CustomLineElement |
    CustomActionLinkElement |
    CustomFeatureLinkElement |
    CustomParagraphElement |
    CustomDescriptionElement



declare module 'slate' {
    interface CustomTypes {
      Editor: BaseEditor & ReactEditor & { saveSelection?: Selection }
      Element: CustomElement
      Text: CustomText
    }
  }

export {}