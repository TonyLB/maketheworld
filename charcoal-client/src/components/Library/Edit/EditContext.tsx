import React, { FunctionComponent, useContext } from "react"
import { GenericTreeNode } from "@tonylb/mtw-wml/dist/tree/baseClasses";
import { SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { TreeId } from "@tonylb/mtw-wml/dist/tree/baseClasses";

type EditContextType = {
    field: GenericTreeNode<SchemaTag, TreeId>;
    parentId: string;
    tag: 'ShortName' | 'Name' | 'Description' | 'Statement' | 'Fallthrough' | 'If'
}

const EditContext = React.createContext<EditContextType>({
    field: { data: { tag: 'Name' }, id: '', children: [] },
    parentId: '',
    tag: 'Name'
})

export const EditSchema: FunctionComponent<EditContextType> = ({ field, parentId, tag, children }) => {
    return <EditContext.Provider value={{ field, parentId, tag }}>
        { children }
    </EditContext.Provider>
}

export const useEditContext = () => (useContext(EditContext))
