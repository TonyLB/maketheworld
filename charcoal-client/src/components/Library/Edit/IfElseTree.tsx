import Box from "@mui/material/Box"
import Chip from "@mui/material/Chip"
import { blue } from "@mui/material/colors"
import React, { FunctionComponent, useCallback, ReactChild, ReactChildren, ReactElement, useMemo } from "react"
import CodeEditor from "./CodeEditor"
import { LabelledIndentBox } from "./LabelledIndentBox"

import AddIcon from '@mui/icons-material/Add'
import ExitIcon from '@mui/icons-material/CallMade'

import { useLibraryAsset } from "./LibraryAsset"
import { SchemaConditionFallthroughTag, SchemaConditionStatementTag, isSchemaCondition, isSchemaConditionFallthrough, isSchemaConditionStatement } from "@tonylb/mtw-wml/dist/schema/baseClasses"
import { EditSchema, useEditContext } from "./EditContext"
import { Radio } from "@mui/material"
import { treeNodeTypeguard } from "@tonylb/mtw-wml/dist/tree/baseClasses"

const AddConditionalButton: FunctionComponent<{ onClick: () => void; label: string }> = ({ onClick, label }) => {
    const { readonly } = useLibraryAsset()
    const onClickHandler = useCallback(() => {
        if (!readonly) {
            onClick()
        }
    }, [onClick, readonly])
    return <Chip
        variant="filled"
        size="small"
        sx={{
            backgroundColor: blue[50],
            borderStyle: "solid",
            borderWidth: "1px",
            borderColor: blue[300],
            '&:hover': {
                backgroundColor: blue[100]
            }
        }}
        label={`+ ${label}`}
        onClick={onClickHandler}
        disabled={readonly}
    />
}

const AddItemButton: FunctionComponent<{ onClick: () => void, addItemIcon: ReactElement }> = ({ onClick, addItemIcon }) => {
    const { readonly } = useLibraryAsset()
    return <Chip
        sx={{
            "&&:hover": {
                background: blue[200]
            },
            background: blue[50],
            borderColor: blue[500],
            borderWidth: '2px',
            borderStyle: "solid"
        }}
        size="small"
        disabled={readonly}
        onClick={onClick}
        icon={<AddIcon />}
        label={addItemIcon}
    />
}

type IfElseWrapBoxProps = {
    id: string;
    type: 'if' | 'elseIf' | 'else';
    source: string;
    actions: ReactChild[] | ReactChildren;
    onDelete: () => void;
    showSelected?: boolean;
    selected?: boolean;
    onSelect?: (id: string) => void;
}

const IfElseWrapBox: FunctionComponent<IfElseWrapBoxProps> = ({ type, source, id, actions, onDelete, showSelected = false, selected = false, onSelect = () => {}, children }) => {
    const { updateSchema } = useLibraryAsset()
    const onChange = useCallback((event) => {
        if (event.target.value) {
            onSelect(id)
        }
    }, [id, onSelect])
    return <LabelledIndentBox
        color={blue}
        label={
            type === 'else'
                ? <React.Fragment>
                    { showSelected && <Radio
                        value={id}
                        checked={selected}
                        onChange={onChange}
                        inputProps={{ 'aria-label': 'Else selected' }}
                    /> }
                    Else
                </React.Fragment>
                : <React.Fragment>
                    { showSelected && <Radio
                        value={id}
                        checked={selected}
                        onChange={onChange}
                        inputProps={{ 'aria-label': 'If selected' }}
                    /> }
                    { type === 'if' ? 'If' : 'Else If' }
                    <Box
                        sx={{
                            display: 'inline-block',
                            borderRadius: '0.25em',
                            backgroundColor: blue[50],
                            paddingLeft: '0.25em',
                            paddingRight: '0.25em',
                            marginRight: '0.25em',
                            minWidth: '1.5em'
                        }}
                    >
                        <CodeEditor
                            source={source}
                            onChange={(source: string) => {
                                updateSchema({
                                    type: 'updateNode',
                                    id,
                                    item: { tag: 'Statement', if: source }
                                })
                            }}
                        />
                    </Box>
                </React.Fragment>
        }
        actions={actions}
        onDelete={onDelete}
    >
        { children }
    </LabelledIndentBox>
}

type IfElseTreeProps = {
    render: FunctionComponent<{}>;
    showSelected?: boolean;
}

//
// IfElseTree assumes that the EditContext passed will have a single conditional top-level element,
// and renders the statements and fallthrough of its children
//
export const IfElseTree = ({ render: Render, showSelected = false }: IfElseTreeProps): ReactElement => {
    const { field } = useEditContext()
    const { updateSchema } = useLibraryAsset()
    const firstStatement = useMemo(() => (field.children[0]), [field])
    const otherStatements = useMemo(() => (field.children.slice(1)), [field])
    const onSelect = useCallback((id: string) => {
        const currentSelected = field.children
            .filter(treeNodeTypeguard((data): data is SchemaConditionStatementTag | SchemaConditionFallthroughTag => (isSchemaConditionStatement(data) || isSchemaConditionFallthrough(data))))
            .find(({ data }) => (data.selected))
        const toSelect = field.children
            .filter(treeNodeTypeguard((data): data is SchemaConditionStatementTag | SchemaConditionFallthroughTag => (isSchemaConditionStatement(data) || isSchemaConditionFallthrough(data))))
            .find((node) => (node.id === id))
        if (currentSelected?.id === toSelect?.id) {
            return
        }
        if (currentSelected) {
            updateSchema({ type: 'updateNode', id: currentSelected.id, item: { ...currentSelected.data, selected: false } })
        }
        if (toSelect) {
            updateSchema({ type: 'updateNode', id: toSelect.id, item: { ...toSelect.data, selected: true } })
        }
    }, [field])
    const addElseIf = useCallback((afterId: string) => (
        <AddItemButton
            key="elseIf"
            addItemIcon={<React.Fragment>elseIf</React.Fragment>}
            onClick={() => {
                updateSchema({ type: 'addChild', id: field.id, afterId, item: { data: { tag: 'Statement', if: '' }, children: [{ data: { tag: 'String', value: '' }, children: [] }]
            } })}}
        />), [field, firstStatement, updateSchema])
    const addElse = useMemo(() => (<AddItemButton key="else" addItemIcon={<React.Fragment>else</React.Fragment>} onClick={() => { updateSchema({ type: 'addChild', id: field.id, item: { data: { tag: 'Fallthrough' }, children: [{ data: { tag: 'String', value: '' }, children: [] }] }})}} />), [field, otherStatements, updateSchema])
    if (
        !(isSchemaCondition(field.data)) ||
        field.children.length === 0 ||
        !isSchemaConditionStatement(field.children[0].data) ||
        Boolean(field.children.find(({ data }) => (!(isSchemaConditionStatement(data) || isSchemaConditionFallthrough(data)))))
    ) {
        throw new Error('Invalid arguments in IfElseTree')
    }
    if (!isSchemaConditionStatement(firstStatement.data)) {
        throw new Error('Invalid arguments in IfElseTree')
    }
    return <React.Fragment>
        <IfElseWrapBox
            key={firstStatement.id}
            id={firstStatement.id}
            type={'if'}
            source={firstStatement.data.if}
            onDelete={() => {
                if (otherStatements.length && isSchemaConditionStatement(otherStatements[0].data)) {
                    updateSchema({ type: 'delete', id: firstStatement.id })
                }
            }}
            actions={[
                addElseIf(firstStatement.id),
                ...(otherStatements.length === 0 ? [addElse] : [])
            ]}
            showSelected={showSelected}
            selected={firstStatement.data.selected}
            onSelect={onSelect}
        >
            <EditSchema tag="Statement" field={firstStatement} parentId={field.id}>
                <Render />
            </EditSchema>
        </IfElseWrapBox>
        { 
            otherStatements.map(({ data, children, id }, index) => {
                return isSchemaConditionStatement(data)
                    ? <IfElseWrapBox
                        key={id}
                        id={id}
                        type={'elseIf'}
                        source={data.if}
                        onDelete={() => { updateSchema({ type: 'delete', id })}}
                        actions={[
                            addElseIf(id),
                            ...(index === otherStatements.length - 1 ? [addElse] : [])
                        ]}
                        showSelected={showSelected}
                        selected={data.selected}
                        onSelect={onSelect}
                    >
                        <EditSchema tag="Statement" field={{ data, children, id }} parentId={field.id}>
                            <Render />
                        </EditSchema>
                    </IfElseWrapBox>
                    : isSchemaConditionFallthrough(data)
                        ? <IfElseWrapBox
                            key={id}
                            id={id}
                            type={'else'}
                            source=''
                            onDelete={() => { updateSchema({ type: 'delete', id })}}
                            actions={[]}
                            showSelected={showSelected}
                            selected={data.selected}
                            onSelect={onSelect}
                        >
                            <EditSchema tag="Fallthrough" field={{ data, children, id }} parentId={field.id}>
                                <Render />
                            </EditSchema>
                        </IfElseWrapBox>
                        : null
            })
        }
    </React.Fragment>
}

export default IfElseTree
