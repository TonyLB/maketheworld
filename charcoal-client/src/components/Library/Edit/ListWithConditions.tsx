import React, { FunctionComponent, PropsWithChildren, ReactElement } from "react"

import { isSchemaCondition, SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses"
import { GenericTreeNodeFiltered, treeNodeTypeguard } from "@tonylb/mtw-wml/dist/tree/baseClasses"
import { EditSubListSchema, useEditContext } from "./EditContext";
import IfElseTree from "./IfElseTree";

type ListWithConditionsProperties<T extends SchemaTag> = {
    typeGuard: (value: SchemaTag) => value is T;
    render: FunctionComponent<{ item: GenericTreeNodeFiltered<T, SchemaTag>, index: number }>;
}

export const ListWithConditions = <T extends SchemaTag>(props: PropsWithChildren<ListWithConditionsProperties<T>>): ReactElement<any, any> | null => {
    const { render: Render } = props
    const { value } = useEditContext()

    return <React.Fragment>{
        value.map((item, index) => (
            <EditSubListSchema
                key={`list-with-conditions-${index}`}
                index={index}
            >
                {
                    treeNodeTypeguard(isSchemaCondition)(item)
                        ? <IfElseTree render={props.render} />
                        : treeNodeTypeguard(props.typeGuard)(item)
                            ? <Render item={item} index={index} />
                            : null
                }
            </EditSubListSchema>
        ))
    }</React.Fragment>
}

export default ListWithConditions
