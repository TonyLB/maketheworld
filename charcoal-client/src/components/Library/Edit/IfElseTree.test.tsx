jest.mock('./CodeEditor')
import CodeEditor from './CodeEditor'

import renderer from 'react-test-renderer'
import React, { FunctionComponent } from 'react'

import ListWithConditions from './ListWithConditions'
import { isSchemaExit, isSchemaOutputTag, SchemaExitTag, SchemaTag } from '@tonylb/mtw-wml/dist/schema/baseClasses'
import { GenericTreeNodeFiltered, treeNodeTypeguard } from '@tonylb/mtw-wml/dist/tree/baseClasses'
import { schemaOutputToString } from '@tonylb/mtw-wml/dist/schema/utils/schemaOutput/schemaOutputToString'
import { EditSchema, useEditContext } from './EditContext'
import { treeTypeGuard } from '@tonylb/mtw-wml/dist/tree/filter'
import IfElseTree from './IfElseTree'

describe('IfElseTree component', () => {
    const render: FunctionComponent<{}> = () => {
        const { value } = useEditContext()
        return <React.Fragment>
            { value
                .filter(treeNodeTypeguard(isSchemaExit))
                .map(({ data, children }) => (<div>{ `'${data.from}' to '${data.to}': ${schemaOutputToString(treeTypeGuard({ typeGuard: isSchemaOutputTag, tree: children }))}` }</div>))
            }
        </React.Fragment>
    }

    beforeEach(() => {
        (CodeEditor as jest.Mock).mockReturnValue(null)
    })

    it('renders single statement correctly', () => {
        expect(renderer
            .create(
                <EditSchema
                    field={{ data: { tag: 'String', value: '' } ,children: [], id: '' }}
                    value={[{
                        data: { tag: 'If' },
                        children: [
                            {
                                data: { tag: 'Statement', if: 'true' },
                                children: [
                                    { data: { tag: 'Exit', from: 'room1', to: 'room2', key: 'room1#room2' }, children: [{ data: { tag: 'String', value: 'closet' }, children: [] }]},
                                    { data: { tag: 'Exit', from: 'room1', to: 'room3', key: 'room1#room3' }, children: [{ data: { tag: 'String', value: 'lobby' }, children: [] }]}
                                ]
                            }
                        ]
                    }]}
                    onChange={() => {}}
                >
                    <IfElseTree render={render} />
                </EditSchema>
            ).toJSON()
        ).toMatchSnapshot()
    })

})