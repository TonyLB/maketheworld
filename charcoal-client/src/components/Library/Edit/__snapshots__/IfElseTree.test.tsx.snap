// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IfElseTree component renders else correctly 1`] = `
Array [
  <div
    className="MuiBox-root css-n8b9os"
  >
    <div
      className="MuiBox-root css-15r38cd"
    >
      If
      <div
        className="MuiBox-root css-1jbq9h"
      />
    </div>
    <div
      className="MuiBox-root css-90szgu"
      onClick={[Function]}
    >
      <span>
        <div>
          'room1' to 'room2': closet
        </div>
      </span>
    </div>
    <div
      className="MuiBox-root css-1lzl05c"
    >
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          elseIf
        </span>
      </div>
    </div>
    <div
      className="MuiBox-root css-klha72"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>,
  <div
    className="MuiBox-root css-n8b9os"
  >
    <div
      className="MuiBox-root css-15r38cd"
    >
      Else
    </div>
    <div
      className="MuiBox-root css-90szgu"
      onClick={[Function]}
    >
      <span>
        <div>
          'room1' to 'room3': lobby
        </div>
      </span>
    </div>
    <div
      className="MuiBox-root css-1lzl05c"
    />
    <div
      className="MuiBox-root css-klha72"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;

exports[`IfElseTree component renders elseif correctly 1`] = `
Array [
  <div
    className="MuiBox-root css-n8b9os"
  >
    <div
      className="MuiBox-root css-15r38cd"
    >
      If
      <div
        className="MuiBox-root css-1jbq9h"
      />
    </div>
    <div
      className="MuiBox-root css-90szgu"
      onClick={[Function]}
    >
      <span>
        <div>
          'room1' to 'room2': closet
        </div>
      </span>
    </div>
    <div
      className="MuiBox-root css-1lzl05c"
    >
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          elseIf
        </span>
      </div>
    </div>
    <div
      className="MuiBox-root css-klha72"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  <div
    className="MuiBox-root css-n8b9os"
  >
    <div
      className="MuiBox-root css-15r38cd"
    >
      Else If
      <div
        className="MuiBox-root css-1jbq9h"
      />
    </div>
    <div
      className="MuiBox-root css-90szgu"
      onClick={[Function]}
    >
      <span>
        <div>
          'room1' to 'room3': lobby
        </div>
      </span>
    </div>
    <div
      className="MuiBox-root css-1lzl05c"
    >
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          elseIf
        </span>
      </div>
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          else
        </span>
      </div>
    </div>
    <div
      className="MuiBox-root css-klha72"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
]
`;

exports[`IfElseTree component renders single statement correctly 1`] = `
<div
  className="MuiBox-root css-n8b9os"
>
  <div
    className="MuiBox-root css-15r38cd"
  >
    If
    <div
      className="MuiBox-root css-1jbq9h"
    />
  </div>
  <div
    className="MuiBox-root css-90szgu"
    onClick={[Function]}
  >
    <span>
      <div>
        'room1' to 'room2': closet
      </div>
      <div>
        'room1' to 'room3': lobby
      </div>
    </span>
  </div>
  <div
    className="MuiBox-root css-1lzl05c"
  >
    <div
      aria-disabled={true}
      className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={-1}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        elseIf
      </span>
    </div>
    <div
      aria-disabled={true}
      className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={-1}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        else
      </span>
    </div>
  </div>
  <div
    className="MuiBox-root css-klha72"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </div>
</div>
`;
